### PR TITLE
Add support for Mutual-TLS Remote requests

### PR DESCRIPTION
**Pull request by fhossain**
This pull request is in response to issue #611 (Mutual auth for remote requests) @lziest

### CODE REVIEW

**Comment by kisom** [comment]
@fhossain Can you run [`golint`](https://github.com/golang/lint) locally? It looks like golint is complaining about some of the changes.

**Comment by fhossain** [comment]
@kisom I fixed the golint errors but there still seems to be a ci failure, is that a known issue or did my change cause it?

File: api/client/client.go [hunk 1]
```diff
@@ -24,6 +26,8 @@ import (
 // A server points to a single remote CFSSL instance.
 type server struct {
 	URL string
+	RemoteCAs *x509.CertPool
```
Comment (jacobhaven) [inline api/client/client.go]: I'd prefer to just pass a full `*tls.Config` around, with the `Certificates` and `RootCAs` contained.

File: api/client/client.go [hunk 2]
```diff
@@ -38,23 +42,24 @@ type Remote interface {
...
-// NewServer sets up a new server target. The address should be the
-// DNS name (or "name:port") of the remote CFSSL instance. If no port
+// NewServer sets up a new server target. The address should be of
+// The format [protocol:]name[:port] of the remote CFSSL instance.
+// If no protocol is given http is default. If no port
 // is specified, the CFSSL default port (8888) is used. If the name is
 // a comma-separated list of hosts, an ordered group will be returned.
-func NewServer(addr string) Remote {
-	addrs := strings.Split(addr, ",")
+func NewServer(addr string, remoteCAs *x509.CertPool, cert *tls.Certificate) Remote {
```
Comment (jacobhaven) [inline api/client/client.go]: I might just add `NewServeTLSr` and `NewGroupTLS` and keep the old API the same signature. `NewServer` is then just:

```Go
func NewServer(addr string) Remote {
	return NewServerTLS(addr, nil)
}
```

File: cli/info/info.go [hunk 3]
```diff
@@ -32,8 +32,15 @@ func getInfoFromRemote(c cli.Config) (resp *info.Resp, err error) {
...
-	serv := client.NewServer(c.Remote)
-
+    cert, err := helpers.LoadClientCertificate(c.MutualTLSCertFile, c.MutualTLSKeyFile)
+    if err != nil {
+		return
+	}
+    remoteCAs, err := helpers.LoadPEMCertPool(c.TLSRemoteCAs)
+    if err != nil {
+		return
+	}
```
Comment (jacobhaven) [inline cli/info/info.go]: I would costruct a `*tls.Config` here and pass that instead.

```Go
serv := client.NewServerTLS(c.Remote, &tls.Config{RootCAs: remoteCAs, Certificates: []tls.Certificate{cert}})
```

**Comment by jacobhaven** [comment]
Your CI error is still a golint/go fmt problem. Mixing spaces and tabs, incorrect indentation, etc. This whitespace diff should fix them all:

```
diff --git a/api/client/client.go b/api/client/client.go
index 678dd6d..305b3fc 100644
--- a/api/client/client.go
+++ b/api/client/client.go
@@ -25,9 +25,9 @@ import (
 
 // A server points to a single remote CFSSL instance.
 type server struct {
-	URL string
+	URL       string
 	RemoteCAs *x509.CertPool
-	Cert *tls.Certificate
+	Cert      *tls.Certificate
 }
 
 // A Remote points to at least one (but possibly multiple) remote
@@ -48,7 +48,7 @@ type Remote interface {
 // is specified, the CFSSL default port (8888) is used. If the name is
 // a comma-separated list of hosts, an ordered group will be returned.
 func NewServer(addr string, remoteCAs *x509.CertPool, cert *tls.Certificate) Remote {
-    addrs := strings.Split(addr, ",")
+	addrs := strings.Split(addr, ",")
 
 	var remote Remote
 
@@ -84,14 +84,14 @@ func (srv *server) createTransport() (transport *http.Transport) {
 	// Setup HTTPS client
 	var certs []tls.Certificate
 	if srv.Cert != nil {
-	    certs = []tls.Certificate{*srv.Cert}
+		certs = []tls.Certificate{*srv.Cert}
 	}
-    tlsConfig := &tls.Config {
-        Certificates: certs,
-        RootCAs: srv.RemoteCAs,
-    }
-    tlsConfig.BuildNameToCertificate()
-    return &http.Transport{TLSClientConfig: tlsConfig}
+	tlsConfig := &tls.Config{
+		Certificates: certs,
+		RootCAs:      srv.RemoteCAs,
+	}
+	tlsConfig.BuildNameToCertificate()
+	return &http.Transport{TLSClientConfig: tlsConfig}
 }
 
 // post connects to the remote server and returns a Response struct
@@ -100,13 +100,13 @@ func (srv *server) post(url string, jsonData []byte) (*api.Response, error) {
 	var resp *http.Response
 	var err error
 	if srv.Cert != nil || srv.RemoteCAs != nil {
-	    transport := srv.createTransport();
-        client := &http.Client{Transport: transport}
-        resp, err = client.Post(url, "application/json", buf)
-    } else {
-        resp, err = http.Post(url, "application/json", buf)
-    }
-    if err != nil {
+		transport := srv.createTransport()
+		client := &http.Client{Transport: transport}
+		resp, err = client.Post(url, "application/json", buf)
+	} else {
+		resp, err = http.Post(url, "application/json", buf)
+	}
+	if err != nil {
 		return nil, errors.Wrap(errors.APIClientError, errors.ClientHTTPError, err)
 	}
 	body, err := ioutil.ReadAll(resp.Body)
diff --git a/api/client/client_test.go b/api/client/client_test.go
index 3f1bd2c..1e4f592 100644
--- a/api/client/client_test.go
+++ b/api/client/client_test.go
@@ -1,7 +1,7 @@
 package client
 
 import (
-    "crypto/tls"
+	"crypto/tls"
 	"github.com/cloudflare/cfssl/auth"
 	"github.com/cloudflare/cfssl/helpers"
 	"net"
@@ -88,7 +88,7 @@ func TestSign(t *testing.T) {
 }
 
 func TestNewMutualTLSServer(t *testing.T) {
-    cert, _ := helpers.LoadClientCertificate("../../helpers/testdata/ca.pem", "../../helpers/testdata/ca_key.pem")
+	cert, _ := helpers.LoadClientCertificate("../../helpers/testdata/ca.pem", "../../helpers/testdata/ca_key.pem")
 	s := NewServer("https://nohost:8888", nil, cert)
 	if s == nil {
 		t.Fatalf("fatal error, empty server")
@@ -98,7 +98,7 @@ func TestNewMutualTLSServer(t *testing.T) {
 		t.Fatalf("expected error with sign function")
 	}
 	if !strings.Contains(err.Error(), "Post https://nohost:8888/api/v1/cfssl/sign: dial tcp: lookup nohost: no such host") {
-	    t.Fatalf("no error message %v", err)
+		t.Fatalf("no error message %v", err)
 	}
 }
 
@@ -145,8 +145,8 @@ func TestNewTLSServerGroup(t *testing.T) {
 }
 
 func TestNewMutualTLSServerGroup(t *testing.T) {
-    cert, _ := helpers.LoadClientCertificate("../../helpers/testdata/ca.pem", "../../helpers/testdata/ca_key.pem")
-    NewTLSServerGroup(t, cert)
+	cert, _ := helpers.LoadClientCertificate("../../helpers/testdata/ca.pem", "../../helpers/testdata/ca_key.pem")
+	NewTLSServerGroup(t, cert)
 }
 
 func NewTLSServerGroup(t *testing.T, cert *tls.Certificate) {
diff --git a/api/client/group.go b/api/client/group.go
index fc87085..be63a14 100644
--- a/api/client/group.go
+++ b/api/client/group.go
@@ -1,8 +1,8 @@
 package client
 
 import (
-    "crypto/tls"
-    "crypto/x509"
+	"crypto/tls"
+	"crypto/x509"
 	"errors"
 	"strings"
```

**Comment by fhossain** [comment]
Thanks @jacobhaven for your feedback. I will address these shortly. The tab vs space was really bugging me. I will have to check my IDE setting for go and fix them for the future as well.

File: api/client/client.go (hunk 1 above)
Comment (fhossain) [inline api/client/client.go]: Fixed.

File: api/client/client.go (hunk 2 above)
Comment (fhossain) [inline api/client/client.go]: Fixed.

File: cli/info/info.go (hunk 3 above)
Comment (fhossain) [inline cli/info/info.go]: Fixed.

**Comment by fhossain** [comment]
@kisom @jacobhaven the commit should have had a clean ci run but somehow the Mac VM run seems to have hung due to unrelated reasons. Can either of you kick off a Travis rerun? Thanks.

**Comment by kisom** [comment]
@fhossain I kicked it again, but it still hangs, which indicates something in this PR is affecting it. The packages that are missing from the tested list are

+ github.com/cloudflare/cfssl/signer/local
+ github.com/cloudflare/cfssl/signer/remote
+ github.com/cloudflare/cfssl/signer/universal
+ github.com/cloudflare/cfssl/transport
+ github.com/cloudflare/cfssl/transport/ca/localca
+ github.com/cloudflare/cfssl/transport/kp
+ github.com/cloudflare/cfssl/ubiquity
+ github.com/cloudflare/cfssl/whitelist

**Comment by fhossain** [comment]
@kisom the problem is in test github.com/cloudflare/cfssl/signer/local
There the test TestCTFailure is hitting an error of the form -
2016/06/06 17:51:33 Got Post https://ct.googleapis.com/pilot/ct/v1/add-pre-chain: stream error: stream ID 1; PROTOCOL_ERROR, backing off.
2016/06/06 17:51:33 Got Unknown, backing-off 10 seconds.
This error keeps on repeating every 10 seconds but at times the requests pass. In the builds that do pass I can see that this test can run for 100's of seconds. That means it has had at least 10 failures before it passes. Here is what I have seen -
1) If I rollback to a master branch, the error problem still exists but the passes after 2/3 failures. If I run on my pull branch it hits more failures before it passes
2) If I change endpoint to http://ct.googleapis.com/pilot instead of https://ct.googleapis.com/pilot then everything is good and the test passes in first attempt
3) I have run a TCP dump while having the failures. It seems like after a failure in github.com/google/certificate-transparency/go/client/logclient.go the code retries the request. However, the new attempt continues over a previous failed SSL connection and succeeds. That seems a bit odd since I would have expected a reattempt would create a completely new SSL connection. I will look into this further and see if I can correlate this to my pull request somehow. 
Do we have any example of CI tests passing in the last one or two days that I can compare the results with. I want to rule out that this is not a recent problem of https://ct.googleapis.com/pilot endpoint.

**Comment by fhossain** [comment]
@kisom this hang seems to be a problem with google/certificate-transparency and not related to my change. This will cause random failures on the master as well. It seems like they also hit the problem but they only cleaned up from fetchAndParse() but not postAndParse()
https://github.com/google/certificate-transparency/commit/a600f7060ec7eb526e1a3f390b0d554f78293df1
I have confirmed that removing the httpReq.Header.Set("Keep-Alive", "timeout=15, max=100") from postAndParse() does indeed fix the problem. I will ask folks in the other repo to clean that up. For now a workaround is to use http to access google api. Is that an acceptable solution?

**Comment by fhossain** [comment]
@kisom we have a clean build now.

**Comment by kisom** [comment]
@fhossain Currently reviewing, but it might take a while as it's a large PR.

**Comment by kisom** [comment]
LGTM, except there's no additions to the documentation for this. Can you document this?

**Comment by fhossain** [comment]
@kisom I would be happy to add the docs but I am not sure where to add it. I think it would make sense to add them to the same place where other tls related parameters (-tls-cert and -tls-key) are documented but I could not locate them. Maybe we need to add them too.

**Comment by kisom** [comment]
@fhossain You're right, it doesn't appear that CLI documentation appears outside the cli/*/*.go files. I'll file a separate issue for that.

@lziest @grittygrease Thoughts on this PR?

File: cli/info/info.go [hunk 4]
```diff
@@ -32,8 +32,15 @@ func getInfoFromRemote(c cli.Config) (resp *info.Resp, err error) {
 	req.Label = c.Label
 	req.Profile = c.Profile
 
-	serv := client.NewServer(c.Remote)
-
+	cert, err := helpers.LoadClientCertificate(c.MutualTLSCertFile, c.MutualTLSKeyFile)
```
Comment (lziest) [inline cli/info/info.go]: with default "" certFile and keyFile, this always errs out? May need some protection.

File: cli/info/info.go [hunk 5]
```diff
@@ -32,8 +32,15 @@ func getInfoFromRemote(c cli.Config) (resp *info.Resp, err error) {
...
+	if err != nil {
+		return
+	}
+	remoteCAs, err := helpers.LoadPEMCertPool(c.TLSRemoteCAs)
```
Comment (lziest) [inline cli/info/info.go]: ditto.

Comment (lziest) [inline cli/info/info.go]: Ignore the above comments. I see you update the behavior of helper functions.

File: helpers/helpers.go [hunk 6]
```diff
@@ -311,11 +312,23 @@ func ParseOneCertificateFromPEM(certsPEM []byte) ([]*x509.Certificate, []byte, e
...
 	if err != nil {
 		return nil, err
 	}
 
+	return PEMToCertPool(pemCerts)
+}
+
+// PEMToCertPool concerts PEM certificates to a CertPool.
+func PEMToCertPool(pemCerts []byte) (*x509.CertPool, error) {
+	if pemCerts == nil {
```
Comment (lziest) [inline helpers/helpers.go]: use len(pemCerts)==0 ?

File: config/config.go [hunk 7]
```diff
@@ -303,6 +306,44 @@ func (p *Signing) OverrideRemotes(remote string) error {
 	return nil
 }
 
+// SetClientCertAndKey updates the properties to set client certificates for mutual
+// authenticated TLS remote requests
+func (p *Signing) SetClientCertAndKey(certFile string, keyFile string) error {
```
Comment (lziest) [inline config/config.go]: naming, can we call it "SetClientKeyPairFromFile"

File: helpers/helpers_test.go [hunk 8]
```diff
@@ -422,3 +422,31 @@ func TestParseOldCSR(t *testing.T) {
 		t.Fatalf("%v", err)
 	}
 }
+
```
Comment (lziest) [inline helpers/helpers_test.go]: can I have some simple tests on the new behavior of LoadPEMCertPool? empty path and empty file.

File: helpers/helpers_test.go [hunk 9]
```diff
@@ -422,3 +422,31 @@ func TestParseOldCSR(t *testing.T) {
...
+// Imported from signers/local/testdata/
+const clientCertFile = "testdata/ca.pem"
+const clientKeyFile = "testdata/ca_key.pem"
+
+func TestClientCertParams(t *testing.T) {
```
Comment (lziest) [inline helpers/helpers_test.go]: Test LoadClientCertificate("", "") as well.

File: helpers/helpers.go (hunk 6 above)
Comment (fhossain) [inline helpers/helpers.go]: Done

File: config/config.go (hunk 7 above)
Comment (fhossain) [inline config/config.go]: Done.

File: helpers/helpers_test.go (hunk 8 above)
Comment (fhossain) [inline helpers/helpers_test.go]: Done

File: helpers/helpers_test.go (hunk 9 above)
Comment (fhossain) [inline helpers/helpers_test.go]: Done

**Comment by lziest** [comment]
LGTM, please squash commits :)

**Comment by fhossain** [comment]
@kisom what's the next step?

**Comment by kisom** [comment]
@fhossain Can you squash the commits?

**Comment by kisom** [comment]
@fhossain Once that's done, we can merge this.

**Comment by fhossain** [comment]
@kisom I thought you could just do a squash merge. Let me know if that is not an option and I can do a rebase.

**Comment by kisom** [comment]
@fhossain Forgot that Github lets you do that now. Thanks!

**Pull request merged**